mlir/lib/Quantum/Transforms/AdjointPatterns.cpp

```diff
@@ -0,0 +1,224 @@
+                    });
+                    auto customA = rewriter.create<CustomOp>(
+                        loc, custom.getResultTypes(), in_params, in_qubits, custom.getGateName(),
+                        mlir::BoolAttr::get(ctx, !custom.getAdjoint().value_or(false)));
+                    for (size_t i = 0; i < customA.getOutQubits().size(); i++) {
+                        update(custom.getQubitOperands()[i], customA->getResult(i));
+                    }
+                }
+                else if (QubitUnitaryOp qunitary = isInstanceOf<QubitUnitaryOp>(*i)) {
+                    assert(
+                        qunitary.getQubitOperands().size() == qunitary.getQubitResults().size() &&
+                        "Quantum operation must have inputs and outputs of the same qubit number");
+                    auto in_qubits = ({
+                        std::vector<Value> qbits;
+                        for (auto q : qunitary.getQubitResults()) {
+                            qbits.push_back(query(q));
+                        }
+                        qbits;
+                    });
+                    auto qunitaryA = rewriter.create<QubitUnitaryOp>(
+                        loc, qunitary.getResultTypes(),
+                        classicalMapping.lookupOrDefault(qunitary.getMatrix()), in_qubits,
+                        mlir::BoolAttr::get(ctx, !qunitary.getAdjoint().value_or(false)));
+                    for (size_t i = 0; i < qunitaryA.getQubitResults().size(); i++) {
+                        update(qunitary.getQubitOperands()[i], qunitaryA->getResult(i));
+                    }
+                }
+                else if (ExtractOp extract = isInstanceOf<ExtractOp>(*i)) {
+                    auto insert = rewriter.create<InsertOp>(
+                        loc, extract.getQreg().getType(), query(extract.getQreg()),
+                        classicalMapping.lookupOrDefault(extract.getIdx()),
+                        extract.getIdxAttrAttr(), query(extract.getQubit()));
+                    update(extract.getQreg(), insert->getResult(0));
+                }
+                else if (AdjointOp adjoint2 = isInstanceOf<AdjointOp>(*i)) {
+                    IRMapping bvm(classicalMapping);
+                    assert(adjoint2.getRegion().hasOneBlock());
+                    Block &b = adjoint2.getRegion().front();
+                    for (const auto &[a, r] : llvm::zip(b.getArguments(), adjoint2->getResults())) {
+                        bvm.map(a, query(r));
+                    }
+                    auto res = copyAdjointVerbatim(adjoint2, rewriter, bvm);
+                    update(adjoint2.getQreg(), res);
+                }
+                else {
+                    /* TODO: We expect to handle Scf control flow instructions here. Stateless loops
+                     * and conditionals should not be a problem. Stateful loops would probably
+                     * require some kind of classical unrolling.
+                     */
```

Comment:
Here the compiler is expected to reverse control flow, is that correct? Any other thing that might be missing here? I just want to understand what is required here.

Reply:
Yes, the generic idea is to recursively reverse the quantum control flow of the scf regions. But we might need to change other places in code, not just this one. I can think of cases where we need to split control flows in order to do the quantum reversal, and of cases where we need to do the unrolling. All because there are classical computations that are not reversible.